services/server/src/server/services/utils/proxy-contract-util.ts

```diff
@@ -0,0 +1,112 @@
+import { whatsabi } from "@shazow/whatsabi";
+import { SourcifyChain } from "@ethereum-sourcify/lib-sourcify";
+import { AbiCoder } from "ethers";
+
+export type ProxyType =
+  | "EIP1167Proxy"
+  | "FixedProxy"
+  | "EIP1967Proxy"
+  | "GnosisSafeProxy"
+  | "DiamondProxy"
+  | "PROXIABLEProxy"
+  | "ZeppelinOSProxy"
+  | "SequenceWalletProxy";
+
+export interface ProxyDetectionResult {
+  isProxy: boolean;
+  proxyType: ProxyType | null;
+  implementations: string[];
+}
+
+export async function detectAndResolveProxy(
+  bytecode: string,
+  address: string,
+  sourcifyChain: SourcifyChain,
+): Promise<ProxyDetectionResult> {
+  // Detection
+  const codeCache = {
+    [address]: bytecode,
+  };
+  const cachedCodeProvider = whatsabi.providers.WithCachedCode(
+    {
+      request: () => {},
+    }, // Provider not needed for detection only
+    codeCache,
+  );
+
+  const detectionResult = await whatsabi.autoload(address, {
+    provider: cachedCodeProvider,
+    abiLoader: false,
+    signatureLookup: false,
+    followProxies: false,
+  });
+  const proxies = detectionResult.proxies;
```

Comment:
For me it's not clear what we are doing after this. We have the `proxies`. Now what? Why are we doing the steps below?

Reply:
I will rename the variable to `proxyResolvers`. The returned object is needed for resolving the implementation address.